parsl/dataflow/dflow.py

```diff
@@ -606,10 +621,13 @@ def checkpoint(self):
             pickle.dump(state, f)
 
         count = 0
+
         with open(checkpoint_tasks, 'ab') as f:
-            for task_id in self.tasks:
-                if self.tasks[task_id]['app_fu'].done() and \
-                   not self.tasks[task_id]['checkpoint']:
+            for task_id in checkpoint_queue:
+
+                if not self.tasks[task_id]['checkpoint'] and \
+                   self.tasks[task_id]['status'] in (States.done, States.failed, States.dep_fail):
```

Comment:
I don't quite follow why we would checkpoint failed tasks? For example, what if there was a transient issue causing tasks to fail, the user stops their script and fixes the issue, then re-runs their script?

Reply:
I think checkpointing failures is the wrong thing to do (in most cases?)

Reply:
Makes sense. There's so much external, non-trackable things that will influence apps.

I'll change this behavior.